### PR TITLE
Script to run kani on top 100 crates

### LINKED ISSUE #1293: Develop a prototype tool to test codegen capabilities for top crates
Create a tool that tests the ability to codegen the 100 top crates in crates.io, some issues we may find:
 1. May not codegen generic functions/types. These would require compiling the tests.
 2. Extract stats on constructs that are not currently supported by Kani as well as unsafe code, etc.
 3. Ensure that it's crates which don't codegen don't pass the test (as the `std` regression).

### PR DESCRIPTION
**Pull request by YoshikiTakashima**
### Description of changes:

Kani itself does not change. This PR adds a script that runs kani on
top 100 crates.

### Resolved issues:

Resolves #1293

### Call-outs:

- Output format is ad-hoc, just whatever grep finds + 3 lines after
  it. If you want a particular format, please request.
- It saids 100 crates, but because of workspaces that have multiple
  top crates, we have 83 repos.
- 34 / 83 of the crates have at least one  warning.
- No change is made to CI.

### Testing:

* How is this change tested? Run the script
  `cat tests/large-crawl/crates-top-100.txt |./scripts/kani-run-on-repos.sh`. Make sure to have enough compute cpu so
  that you don't overload your system. At least 8 threads, 16 is better.

* Is this a refactor change? No.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [x] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by YoshikiTakashima** [comment]
@adpaco-aws Please review when you get back from vacation.

File: scripts/kani-top-100.sh [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+OVERALL_EXIT_CODE='0'
+TARGET_ERROR_REGEX='warning:\sFound\sthe\sfollowing\sunsupported\sconstructs:\|WARN'
+# printing function that greps the error logs and exit code.
+function print_errors_for_each_repo_result {
+    DIRECTORY=$1
+    IS_FAIL='0'
+
+    error_code="$(cat $DIRECTORY/$EXIT_CODE_SUFFIX)"
+    if ! [ "$error_code" = "0" ]; then
+	echo -e "Error exit: code $error_code\n"
```
Comment (fzaiser) [inline scripts/kani-top-100.sh:153]: Indentation seems off. Mix of tabs and spaces?

**Comment by YoshikiTakashima** [comment]
Callouts from Friday's mtg
- JSON will be better.
- Since this finishes running in under 1 minute. Daniel suggests
  expanding the script.
- Ted warned that we may have missed some warnings because
  `--workspace` is not implemented and some may not have gotten
  compiled.

File: scripts/kani-top-100.sh [hunk 2]
```diff
@@ -34,89 +34,7 @@ SELF_SCRIPT=$0
...
-https://github.com/tokio-rs/bytes
-https://github.com/tokio-rs/mio
-https://github.com/tokio-rs/slab
-https://github.com/tokio-rs/tokio
-https://github.com/unicode-rs/unicode-normalization
-https://github.com/unicode-rs/unicode-segmentation
-https://github.com/unicode-rs/unicode-width
-https://github.com/unicode-rs/unicode-xid
-https://github.com/withoutboats/heck'
+HARD_CODED_TOP_100_CRATES_AS_OF_2022_6_17=$(cat $SELF_DIR/../tests/large-crawl/crates-top-100.txt)
```
Comment (celinval) [inline scripts/kani-top-100.sh:37]: Can you make this an argument instead?

Comment (YoshikiTakashima) [reply scripts/kani-top-100.sh]: Argument will be a bit difficult b/c of the way this script calls itself with different arguments during the actual work step. 

I did STDIN instead, but I am willing to do harder modifications if you strongly want arguments.

Comment (YoshikiTakashima) [reply scripts/kani-top-100.sh]: Implemented request in b528facff1604454f2a1326e842e1da1b5b61cab.

File: scripts/kani-run-on-repos.sh [hunk 3]
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
+# Copyright Kani Contributors
+# SPDX-License-Identifier: Apache-2.0 OR MIT
+
+
+DOCUMENTATION=\
+'kani-run-on-repos.sh -- script to clone and compile multiple remote git repositories with Kani.
+
+USAGE:
+./scripts/kani-top-100.sh path/to/url-list
```
Comment (adpaco-aws) [inline scripts/kani-run-on-repos.sh:10]: Script name needs to be updated.

File: scripts/kani-run-on-repos.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
+#!/bin/bash
```
Comment (adpaco-aws) [inline scripts/kani-run-on-repos.sh:1]: I'm wondering if we should create a new `scripts/exps/` folder for these experimental scripts. What do you think?

File: scripts/kani-run-on-repos.sh [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+if ! which xargs 1>&2 1> /dev/null; then
+    echo "Need to have xargs installed. Please install with `apt-get install -y xargs`"
+    exit -1
+elif [[ "$*" == *"--help"* ]]; then
+    echo -e "$DOCUMENTATION"
+elif [ "$#" -eq "1" ]; then
+    # top level logic that runs clone_and_run_kani in parallel with xargs.
+    echo 'Reading URLs from STDIN...';
+    LIST_OF_CRATE_GIT_URLS=$(cat $1)
+    if [ "$LIST_OF_CRATE_GIT_URLS" =~ "\s"* ]; then
```
Comment (adpaco-aws) [inline scripts/kani-run-on-repos.sh:100]: I got this when I run the script:
```
./scripts/kani-run-on-repos.sh: line 100: [: =~: binary operator expected
```
Do you also get it?

**Review by adpaco-aws — COMMENTED** [review]
Thanks @YoshikiTakashima for this!!

I've tried it on my remote machine and I'm getting the following results:
```
--- OVERALL STATS ---
8 crates failed to compile
43 crates had warning(s)
```
How many crates fail to compile in your environment?

There's room for improvement in how error & warnings are collected, but let's leave that for another PR. Also, please see the comments below.

File: scripts/kani-run-on-repos.sh (hunk 5 above)
Comment (YoshikiTakashima) [reply scripts/kani-run-on-repos.sh]: Yes, I think I made a typo with the regex. I changed it to a different emptiness check.

File: scripts/kani-run-on-repos.sh (hunk 3 above)
Comment (YoshikiTakashima) [reply scripts/kani-run-on-repos.sh]: Fixed with 6135a3d8543a50fd54650d972d2450ee23aad63a

File: scripts/kani-run-on-repos.sh (hunk 4 above)
Comment (YoshikiTakashima) [reply scripts/kani-run-on-repos.sh:1]: I don't have strong preference in this area. It depends on whether to consider this as a metric or a test case.

**Comment by YoshikiTakashima** [comment]
Similar number. x86 is a bit better (around 30 warnings) and same number of fails.

File: scripts/kani-run-on-repos.sh (hunk 4 above)
Comment (YoshikiTakashima) [reply scripts/kani-run-on-repos.sh:1]: Also, do we have other scripts than this one to put in such a directory?

Comment (adpaco-aws) [reply scripts/kani-run-on-repos.sh:1]: I think this is going to work mostly as a metric. My concern here is that `scripts` is becoming too cluttered, and I'd propose moving "format" scripts to `ci` and regression scripts to a new folder `tests`. There are no other scripts that would go into this directory, at least for now.

Comment (YoshikiTakashima) [reply scripts/kani-run-on-repos.sh]: I think that is reasonable. Fixed w/ a067c8d6e071652fdad5eefa6efd2e3cf3254734

**Comment by YoshikiTakashima** [comment]
FYI on performance:
- roughly 3 minutes for 100.
- Similar scaling for 1k, so 30 minutes.

**Review by adpaco-aws — APPROVED** [review]

File: tests/remote-target-lists/top-1k-crates-2022-6-27.txt [hunk 6]
```diff
@@ -0,0 +1,621 @@
+http://github.com/SSheldon/rust-block
```
Comment (celinval) [inline tests/remote-target-lists/top-1k-crates-2022-6-27.txt:1]: How hard is it to generate this list? I am tempted to drop this file for now. I can see us integrating the top 100 to our CI, but I'm guessing the top 1k is not in our near future.

File: scripts/exps/kani-run-on-repos.sh [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+OVERALL_EXIT_CODE='0'
+TARGET_ERROR_REGEX='warning:\sFound\sthe\sfollowing\sunsupported\sconstructs:\|WARN'
+# printing function that greps the error logs and exit code.
+function print_errors_for_each_repo_result {
+    DIRECTORY=$1
+    IS_FAIL='0'
+
+    error_code="$(cat $DIRECTORY/$EXIT_CODE_SUFFIX)"
+    if [ "$error_code" != "0" ]; then
```
Comment (celinval) [inline scripts/exps/kani-run-on-repos.sh:69]: Can you please print where the logs are?

**Review by celinval — CHANGES_REQUESTED** [review]
Can you please document somewhere in our [DevDocumentation](https://model-checking.github.io/kani/dev-documentation.html) how to generate a new list of top N crates in crates.io?

File: tests/remote-target-lists/top-1k-crates-2022-6-27.txt (hunk 6 above)
Comment (YoshikiTakashima) [reply tests/remote-target-lists/top-1k-crates-2022-6-27.txt]: OK. Removed with a663ec603c2dbcdcfaa096dade0313ec3d74dcf6

File: scripts/exps/kani-run-on-repos.sh [hunk 8]
```diff
@@ -0,0 +1,137 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (YoshikiTakashima) [reply scripts/exps/kani-run-on-repos.sh:1]: 0064feb6d4bd6ddf8618a0f209d616ce07ecc8e6: Documentation added under `docs/src/repo-crawl.md` on how to derive this list

File: scripts/exps/kani-run-on-repos.sh (hunk 7 above)
Comment (YoshikiTakashima) [reply scripts/exps/kani-run-on-repos.sh:72]: Implemented with f2e5d10c9a250bfb9c3805671108031172295b9f.

**Review by celinval — APPROVED** [review]
Awesome! Thanks you.

Just a small change, can you please add this new page to [SUMMARY.md](https://github.com/model-checking/kani/blob/main/docs/src/SUMMARY.md)?

**Pull request merged**